### PR TITLE
Improve `isatty` check

### PR DESCRIPTION
**Pull request by XiaoliChan**
- use `os.isatty(sys.stdout.fileno())` is much better than `sys.stdout.isatty()`

Fixed:
- #40 
- #41

### CODE REVIEW

**Comment by XiaoliChan** [comment]
> Thanks for the PR, please could you explain how this is better?

The `os.isatty(sys.stdout.fileno())` is do more low-level check than `sys.stdout.isatty()`

**Comment by hugovk** [comment]
Why is the CI failing with `io.UnsupportedOperation: fileno`?

**Comment by XiaoliChan** [comment]
> Why is the CI failing with `io.UnsupportedOperation: fileno`?

Weird

**Comment by XiaoliChan** [comment]
Let me do more test

**Comment by XiaoliChan** [comment]
@hugovk Just update the `test_termcolor.py` to make it pass the check

**Comment by hugovk** [comment]
> > Thanks for the PR, please could you explain how this is better?
> 
> The `os.isatty(sys.stdout.fileno())` is do more low-level check than `sys.stdout.isatty()`

But why is this better?

---

Please could you fix the mypy error?

Also it looks like some other people also check for `io.UnsupportedOperation` - should we do that too?

https://github.com/pdm-project/pdm/blob/99e7877c98bf9beaa8468b1e6ee3365778a63c72/install-pdm.py#L167-L170

**Comment by XiaoliChan** [comment]
> But why is this better?
I can't explain this sorry, but it actually works better than `sys.stdout.isatty()`, so I think it do more deep low-level check.

File: src/termcolor/termcolor.py [hunk 1]
```diff
@@ -120,11 +121,19 @@ def _can_do_colour(
...
+            hasattr(sys.stdout, "fileno")
+            and os.isatty(sys.stdout.fileno())
+            and os.environ.get("TERM") != "dumb"
+        )
+    except io.UnsupportedOperation:
+        return (
+            hasattr(sys.stdout, "isatty")
+            and sys.stdout.isatty()
+            and os.environ.get("TERM") != "dumb"
+        )
```
Comment (hugovk) [inline src/termcolor/termcolor.py:136]: Perhaps something like this to (a) move the duplicate checks out of the `try`/`except` and (b) return early after fewer checks?

```suggestion
    # Then check system:
    if os.environ.get("TERM") == "dumb":
        return False
    if not hasattr(sys.stdout, "fileno"):
        return False

    try:
        return os.isatty(sys.stdout.fileno())
    except io.UnsupportedOperation:
        return sys.stdout.isatty()
```

And is it possible to add a test case to cover the `except` block?

File: tests/test_termcolor.py [hunk 2]
```diff
@@ -25,6 +25,8 @@ def setup_module() -> None:
 def test_basic(monkeypatch: pytest.MonkeyPatch) -> None:
     # Arrange
     monkeypatch.setattr("sys.stdout.isatty", lambda: True)
+    monkeypatch.setattr("sys.stdout.fileno", lambda: 1)
+    monkeypatch.setattr(os, "isatty", lambda fd: True)
```
Comment (hugovk) [inline tests/test_termcolor.py:29]: Nitpick: please could you put these three statements in alphabetical order? And the same below.

File: src/termcolor/termcolor.py (hunk 1 above)
Comment (XiaoliChan) [reply src/termcolor/termcolor.py:136]: This look much better, less duplicate checks

File: tests/test_termcolor.py (hunk 2 above)
Comment (XiaoliChan) [reply tests/test_termcolor.py:29]: Sure!

Comment (XiaoliChan) [reply tests/test_termcolor.py]: Fixed!

**Comment by XiaoliChan** [comment]
Why the workflow stucked here XD

**Comment by hugovk** [comment]
> Why the workflow stucked here XD

I need to approve it for first-time contributors: 

image

**Comment by XiaoliChan** [comment]
> I need to approve it for first-time contributors:

XD

So I think it can be merged, how do you think of it?

**Comment by hugovk** [comment]
Thank you!

**Pull request merged**